File: sortingAlgorithms/InsertionSort/Haskell/InsertionSort.hs
module InsertionSort( sort ) where

-- -----------------------------------------------------------
-- Module       :   InsertionSort.hs
-- Description  :   Implementation of the Insertion sort
-- Copyright    :   (c) Thomas Lang, 2014
-- License      :   BSD3
--
-- Stability    :   stable
-- Portability  :   portable
--
-- This module implements a function that performs the
-- "Insertion sort" algorithm on a list.
-- The insertion sort algorithm compares elements an then
-- it moves that element to it's correct position.
--
-- -> Runs in O(n^2)
-- -----------------------------------------------------------

import Data.List  ( insert )

sort :: (Ord a) => [a] -> [a]
sort = foldr insert []

File: sortingAlgorithms/InsertionSort/Java/InsertionSort.java
import java.util.List;
import java.util.LinkedList;
import java.util.ListIterator;

/**
 * Insertion sort-class.
 *
 * @author Thomas Lang
 * @version 0.2
 */
public class InsertionSort<T extends Comparable<T>>{

    /**
     * Empty Constructor.
     */
    public InsertionSort(){}

    /**
     * Sorting method, performs the insertion
     * sort algorithm.
     */
    public static <T extends Comparable<T>> List<T> insertionSort( T[] array ){
        List<T> list = new LinkedList<T>();
        for( T element : array ){
            ListIterator<T> iterator = list.listIterator(); 
            boolean foundGreater = false;
            while( iterator.hasNext() && !foundGreater )
                foundGreater |= iterator.next().compareTo( element ) > 0;
            if( foundGreater )
                iterator.previous();
            iterator.add( element );
        }
        return list;
    }
}
